README.md
[中文](./README.md)|[English](./README_en.md)



# Telnet

使用 Dart 语言实现的 Telnet 客户端。



## 功能特点

- 支持选项协商、子选项协商和文本消息传输
- 支持 TLS 安全传输
- 消息事件侦听
- 枚举了所有的 Telnet 命令码和选项码



## 使用方法

创建一个 Telnet 连接任务：

```dart
// 常规方式
final task = TelnetClient.startConnect(
  host: host, 
  port: port, 
  timeout: timeout,
  onError: onError,
  onDone: onDone,
  onEvent: onEvent,
);

// 使用 TLS 安全传输
final task = TelnetClient.startSecureConnect(
	host: host,
  port: port,
  timeout: timeout,
  onError: onError,
  onDone: onDone,
  onEvent: onEvent,
  securityContext: securityContext,
  supportedProtocols: supportedProtocols,
  onBadCertificate: onBadCertificate,
);
```

等待连接任务结束，然后获取 `TelnetClient` 实例对象：

```dart
// 同步方式
await task.waitDone();
final client = task.client;
final connected = client != null;

// 异步方式
task.onDone = (client) {
  final connected = client != null;
}
```

取消连接任务：

```dart
task.cancel();
```

关闭 Telnet 连接：

```dart
client.terminate();
```

侦听并处理消息事件：

```dart
final task = TelnetClient.startConnect(
  host: host, 
  port: port, 
  onEvent: (client, event) {
    final eventType = event.type;
    final eventMsg = event.msg;
    
    if (eventType == TLMsgEventType.write) {
      print("这是一个写事件，数据由客户端发往服务端。");
    } else if (eventType == TLMsgEventType.read) {
      print("这是一个读事件，数据由服务端发往客户端。");
    }
    
    if (eventMsg is TLOptMsg) {
      // 选项协商
      print("IAC ${eventMsg.cmd.code} ${eventMsg.opt.code}");
    } else if (eventMsg is TLSubMsg) {
      // 子选项协商
      print("IAC SB ${eventMsg.opt.code} ${eventMsg.arg.join(' ')} IAC SE");
    } else if (eventMsg is TLTextMsg) {
      // 文本消息
      print(eventMsg.text);
    }
  },
);
```

完整使用方法请参考 `Example` 代码，更多详细说明请参考 API 文档。

CHANGELOG.md
## 1.0.7

- Fixed example comment.

## 1.0.6

- Fixed encoding issue.

## 1.0.5

- Update Example Code.

## 1.0.4

- Update Example Code.

## 1.0.3

- Safely terminate the client connection.

## 1.0.2

- Export Interfaces.

## 1.0.1

- Update Document.

## 1.0.0

- Initial version.


test/telnet_test.dart
import 'package:telnet/telnet.dart';
import 'package:test/test.dart';

void main() {
  // 可以通过 docker 创建一个 telnet-server 来进行测试：
  // `docker run -it --name telnet-server -p 60023:23 flemingcsi/telnet-server`
  // 进入容器后执行 `passwd` 修改登录密码。
  const host = "127.0.0.1";
  const port = 60023;
  const username = "root";
  const password = "admin";

  test("Connection and Login", () async {
    var loginSuccessful = false;
    final task = TelnetClient.startConnect(
        host: host,
        port: port,
        onEvent: (client, event) async {
          print(event);
          if (loginSuccessful || event.type != TLMsgEventType.read) {
            return;
          }

          final msg = event.msg;
          if (msg is TLTextMsg) {
            final text = msg.text.toLowerCase();
            if (text.contains("welcome")) {
              loginSuccessful = true;
              client?.write(TLOptMsg(TLCmd.doNot, TLOpt.echo));
              client?.write(TLOptMsg(TLCmd.wont, TLOpt.echo));
              client!.write(TLTextMsg("uname -a\r\n"));
            } else if (text.contains("login:") || text.contains("username:")) {
              client!.write(TLTextMsg("$username\r\n"));
            } else if (text.contains("password:")) {
              client!.write(TLTextMsg("$password\r\n"));
            }
          } else if (msg is TLOptMsg) {
            if (msg.cmd == TLCmd.doIt) {
              client!.write(TLOptMsg(TLCmd.wont, msg.opt));
            } else if (msg.cmd == TLCmd.will) {
              client!.write(TLOptMsg(TLCmd.doNot, msg.opt));
            }
          }
        },
        onError: (cli, err) => print(err));

    await task.waitDone();
    final client = task.client;
    expect(client != null, true);

    await Future.delayed(const Duration(seconds: 10));
    expect(loginSuccessful, true);
    await Future.delayed(const Duration(seconds: 1));
    await task.client?.terminate();
  });

  test("Capture error", () async {
    var hasError = false;
    final task = TelnetClient.startConnect(host: host, port: 99999);
    task.onError = (e) {
      print(e);
      hasError = true;
    };
    await task.waitDone();
    expect(hasError, true);
  });

  test("Cancel connection", () async {
    final task = TelnetClient.startConnect(host: host, port: port);
    task.cancel();
    await task.waitDone();
    expect(task.client, null);
  });

  test("Terminate connection", () async {
    var hasTerminated = false;
    final task = TelnetClient.startConnect(
        host: host,
        port: port,
        onDone: (client) {
          hasTerminated = true;
        });
    await task.waitDone();
    expect(task.client != null, true);
    await Future.delayed(const Duration(seconds: 2));
    expect(hasTerminated, false);
    await task.client?.terminate();
    await Future.delayed(const Duration(seconds: 1));
    expect(hasTerminated, true);
  });

  test("Secure connection", () async {
    var loginSuccessful = false;
    final task = TelnetClient.startSecureConnect(
        host: host,
        port: port,
        onEvent: (client, event) {
          if (loginSuccessful || event.type != TLMsgEventType.read) {
            return;
          }

          final msg = event.msg;
          if (msg is TLTextMsg) {
            final text = msg.text.toLowerCase();
            if (text.contains("welcome")) {
              loginSuccessful = true;
              client?.write(TLOptMsg(TLCmd.doNot, TLOpt.echo));
              client?.write(TLOptMsg(TLCmd.wont, TLOpt.echo));
              client!.write(TLTextMsg("uname -a\r\n"));
            } else if (text.contains("login:") || text.contains("username:")) {
              client!.write(TLTextMsg("$username\r\n"));
            } else if (text.contains("password:")) {
              client!.write(TLTextMsg("$password\r\n"));
            }
          } else if (msg is TLOptMsg) {
            if (msg.cmd == TLCmd.doIt) {
              client!.write(TLOptMsg(TLCmd.wont, msg.opt));
            } else if (msg.cmd == TLCmd.will) {
              client!.write(TLOptMsg(TLCmd.doNot, msg.opt));
            }
          }
        },
        onBadCertificate: (_) => true,
        onError: (client, err) => print(err));

    await task.waitDone();
    expect(task.client != null, true);

    await Future.delayed(const Duration(seconds: 10));
    expect(loginSuccessful, true);

    await Future.delayed(const Duration(seconds: 2));
    await task.client?.terminate();
  });
}


lib/src/code.dart
part of "package:telnet/telnet.dart";

/// Telnet 命令代码。
///
/// 参考资料: [Telnet commands and options](https://www.ibm.com/docs/en/zos/2.3.0?topic=problems-telnet-commands-options)
class TLCmd {
  static final _map = <int, TLCmd>{};

  TLCmd._(this.code) {
    assert(code >= 128 && code <= 255);
    _map[code] = this;
  }

  /// 通过代码数值来获取 [TLCmd] 实例。
  ///
  /// 注意：相同的代码所获得的实例为同一个。
  factory TLCmd.get(int code) {
    return _map[code] ?? TLCmd._(code);
  }

  /// 代码数值。
  final int code;

  @override
  int get hashCode => code;

  @override
  bool operator ==(Object other) => other is TLCmd && other.code == code;

  @override
  String toString() => "$runtimeType($code)";

  /// End of subnegotiation parameters.
  static final se = TLCmd._(240);

  /// No operation.
  static final nop = TLCmd._(241);

  /// Data Mark (part of the Synch function). Indicates the position of a Synch event within the data stream.
  static final dm = TLCmd._(242);

  /// NYT character break.
  static final brk = TLCmd._(243);

  /// Suspend, interrupt or abort the process to which the NVT is connected.
  static final ip = TLCmd._(244);

  /// Abort Output. Allows the current process to run to completion but do not send its output to the user.
  static final ao = TLCmd._(245);

  /// Are you there? AYT is used to determine if the remote TELNET partner is still up and running.
  static final ayt = TLCmd._(246);

  /// Erase character. Erase character is used to indicate the receiver should delete the last preceding undeleted character from the data stream.
  static final ec = TLCmd._(247);

  /// Erase line. Delete characters from the data stream back to but not including the previous CRLF.
  static final el = TLCmd._(248);

  /// Go ahead. Go ahead is used in half-duplex mode to indicate the other end that it can transmit.
  static final ga = TLCmd._(249);

  /// Begin of subnegotiation.
  static final sb = TLCmd._(250);

  /// The sender wants to enable an option.
  ///
  /// 使用场景：
  /// - 发送方想激活选项
  /// - 接收方答复 Do 请求（同意）
  static final will = TLCmd._(251);

  /// The sender do not wants to enable an option.
  ///
  /// 使用场景：
  /// - 发送方想禁止选项
  /// - 接收方答复 Do（不同意）或 Don't（必须同意）请求
  static final wont = TLCmd._(252);

  /// Sender asks receiver to enable an option.
  ///
  /// 使用场景：
  /// - 发送方想让接收方激活选项
  /// - 接收方答复 Will 请求（同意）
  static final doIt = TLCmd._(253);

  /// Sender asks receiver not to enable an option.
  ///
  /// 使用场景：
  /// - 发送方想让接收方禁止选项
  /// - 接收方答复 Will（不同意）或 Won't（必须同意）请求
  static final doNot = TLCmd._(254);

  /// IAC (Interpret as Command).
  static final iac = TLCmd._(255);
}

/// Telnet 选项代码。
///
/// 参考资料: [Telnet Options](https://www.iana.org/assignments/telnet-options/telnet-options.xhtml)
class TLOpt {
  static final _map = <int, TLOpt>{};

  TLOpt._(this.code) {
    _map[code] = this;
  }

  /// 通过代码数值来获取 [TLOpt] 实例。
  ///
  /// 注意：相同的代码所获得的实例为同一个。
  factory TLOpt.get(int code) {
    assert(code >= 0 && code <= 255);
    return _map[code] ?? TLOpt._(code);
  }

  /// 代码数值。
  final int code;

  @override
  int get hashCode => code.hashCode;

  @override
  bool operator ==(Object other) => other is TLOpt && other.code == code;

  @override
  String toString() => "$runtimeType($code)";

  /// Binary Transmission.
  ///
  /// 参考：[RFC856](https://www.rfc-editor.org/rfc/rfc856.html)
  static final bin = TLOpt._(0);

  /// Echo.
  ///
  /// 参考：[RFC857](https://www.rfc-editor.org/rfc/rfc857.html)
  static final echo = TLOpt._(1);

  /// Reconnection.
  ///
  /// 参考：[NIC 15391 of 1973]()
  static final reconn = TLOpt._(2);

  /// Suppress Go Ahead.
  ///
  /// 参考：[RFC858](https://www.rfc-editor.org/rfc/rfc858.html)
  static final suppress = TLOpt._(3);

  /// Approx Message Size Negotiation.
  ///
  /// 参考：[NIC 15393 of 1973]()
  static final msgSize = TLOpt._(4);

  /// Status.
  ///
  /// 参考：[RFC859](https://www.rfc-editor.org/rfc/rfc859.html)
  static final status = TLOpt._(5);

  /// Timing Mark.
  ///
  /// 参考：[RFC860](https://www.rfc-editor.org/rfc/rfc859.html)
  static final timingMark = TLOpt._(6);

  /// Remote Controlled Trans and Echo.
  ///
  /// 参考：[RFC729](https://www.rfc-editor.org/rfc/rfc726.html)
  static final remoteCtrl = TLOpt._(7);

  /// Output Line Width.
  ///
  /// 参考：[NIC 20196 of August 1978]()
  static final oLineWidth = TLOpt._(8);

  /// Output Page Size.
  ///
  /// 参考：[NIC 20197 of August 1978]()
  static final oPageSize = TLOpt._(9);

  /// Output Carriage-Return Disposition.
  ///
  /// 参考：[RFC652](https://www.rfc-editor.org/rfc/rfc652.html)
  static final oCRD = TLOpt._(10);

  /// Output Horizontal Tab Stops.
  ///
  /// 参考：[RFC653](https://www.rfc-editor.org/rfc/rfc653.html)
  static final oHTS = TLOpt._(11);

  /// Output Horizontal Tab Disposition.
  ///
  /// 参考：[RFC654](https://www.rfc-editor.org/rfc/rfc654.html)
  static final oHTD = TLOpt._(12);

  /// Output Formfeed Disposition.
  ///
  /// 参考：[RFC655](https://www.rfc-editor.org/rfc/rfc655.html)
  static final oFD = TLOpt._(13);

  /// 	Output Vertical Tab Stops.
  ///
  /// 参考：[RFC656](https://www.rfc-editor.org/rfc/rfc656.html)
  static final oVTS = TLOpt._(14);

  /// 	Output Vertical Tab Disposition.
  ///
  /// 参考：[RFC657](https://www.rfc-editor.org/rfc/rfc657.html)
  static final oVTD = TLOpt._(15);

  /// Output Linefeed Disposition.
  ///
  /// 参考：[RFC658](https://www.rfc-editor.org/rfc/rfc658.html)
  static final oLD = TLOpt._(16);

  /// Extended ASCII.
  ///
  /// 参考：[RFC698](https://www.rfc-editor.org/rfc/rfc698.html)
  static final extAscii = TLOpt._(17);

  /// Logout.
  ///
  /// 参考：[RFC727](https://www.rfc-editor.org/rfc/rfc727.html)
  static final logout = TLOpt._(18);

  /// Byte Macro.
  ///
  /// 参考：[RFC735](https://www.rfc-editor.org/rfc/rfc735.html)
  static final byteMacro = TLOpt._(19);

  /// Data Entry Terminal.
  ///
  /// 参考：[RFC1043](https://www.rfc-editor.org/rfc/rfc1043.html)、[RFC732](https://www.rfc-editor.org/rfc/rfc732.html)
  static final dataEntry = TLOpt._(20);

  /// SUPDUP.
  ///
  /// 参考：[RFC736](https://www.rfc-editor.org/rfc/rfc736.html)、[RFC734](https://www.rfc-editor.org/rfc/rfc734.html)
  static final supdup = TLOpt._(21);

  /// SUPDUP Output.
  ///
  /// 参考：[RFC749](https://www.rfc-editor.org/rfc/rfc749.html)
  static final supdupOut = TLOpt._(22);

  /// Send Location.
  ///
  /// 参考：[RFC779](https://www.rfc-editor.org/rfc/rfc779.html)
  static final sendLoc = TLOpt._(23);

  /// Terminal Type.
  ///
  /// 参考：[RFC1091](https://www.rfc-editor.org/rfc/rfc1091.html)
  static final tmlType = TLOpt._(24);

  /// End of Record.
  ///
  /// 参考：[RFC885](https://www.rfc-editor.org/rfc/rfc885.html)
  static final eor = TLOpt._(25);

  /// TACACS User Identification.
  ///
  /// 参考：[RFC927](https://www.rfc-editor.org/rfc/rfc927.html)
  static final tacacs = TLOpt._(26);

  /// Output Marking.
  ///
  /// 参考：[RFC933](https://www.rfc-editor.org/rfc/rfc933.html)
  static final oMarking = TLOpt._(27);

  /// Terminal Location Number.
  ///
  /// 参考：[RFC946](https://www.rfc-editor.org/rfc/rfc946.html)
  static final tmlLocNum = TLOpt._(28);

  /// Telnet 3270 Regime.
  ///
  /// 参考：[RFC1041](https://www.rfc-editor.org/rfc/rfc1041.html)
  static final regimeOf3270 = TLOpt._(29);

  /// X.3 PAD.
  ///
  /// 参考：[RFC1053](https://www.rfc-editor.org/rfc/rfc1053.html)
  static final x3Pad = TLOpt._(30);

  /// Negotiate About Window Size.
  ///
  /// 参考：[RFC1073](https://www.rfc-editor.org/rfc/rfc1073.html)
  static final windowSize = TLOpt._(31);

  /// Terminal Speed.
  ///
  /// 参考：[RFC1079](https://www.rfc-editor.org/rfc/rfc1079.html)
  static final tmlSpeed = TLOpt._(32);

  /// Remote Flow Control.
  ///
  /// 参考：[RFC1372](https://www.rfc-editor.org/rfc/rfc1372.html)
  static final flowCtrl = TLOpt._(33);

  /// Line Mode.
  ///
  /// 参考：[RFC1184](https://www.rfc-editor.org/rfc/rfc1184.html)
  static final lineMode = TLOpt._(34);

  /// X Display Location.
  ///
  /// 参考：[RFC1096](https://www.rfc-editor.org/rfc/rfc1096.html)
  static final xDspLoc = TLOpt._(35);

  /// Environment Option.
  ///
  /// 参考：[RFC1408](https://www.rfc-editor.org/rfc/rfc1408.html)
  static final env = TLOpt._(36);

  /// Authentication Option.
  ///
  /// 参考：[RFC2941](https://www.rfc-editor.org/rfc/rfc2941.html)
  static final auth = TLOpt._(37);

  /// Encryption Option.
  ///
  /// 参考：[RFC2946](https://www.rfc-editor.org/rfc/rfc2946.html)
  static final encryption = TLOpt._(38);

  /// New Environment Option.
  ///
  /// 参考：[RFC1572](https://www.rfc-editor.org/rfc/rfc1572.html)
  static final newEnv = TLOpt._(39);

  /// TN3270E.
  ///
  /// 参考：[RFC2355](https://www.rfc-editor.org/rfc/rfc2355.html)
  static final tn3270e = TLOpt._(40);

  /// XAUTH.
  static final xauth = TLOpt._(41);

  /// CHARSET.
  ///
  /// 参考：[RFC2066](https://www.rfc-editor.org/rfc/rfc2066.html)
  static final charset = TLOpt._(42);

  /// Telnet Remote Serial Port (RSP).
  static final rsp = TLOpt._(43);

  /// Com Port Control Option.
  ///
  /// 参考：[RFC2217](https://www.rfc-editor.org/rfc/rfc2217.html)
  static final comPortCtrl = TLOpt._(44);

  /// Telnet Suppress Local Echo.
  static final suppressLocalEcho = TLOpt._(45);

  /// Telnet Start TLS.
  static final startTls = TLOpt._(46);

  /// KERMIT.
  ///
  /// 参考：[RFC2840](https://www.rfc-editor.org/rfc/rfc2840.html)
  static final kermit = TLOpt._(47);

  /// SEND-URL.
  static final sendUrl = TLOpt._(48);

  /// FORWARD_X.
  static final forwardX = TLOpt._(49);

  /// TELOPT PRAGMA LOGON.
  static final pragmaLogon = TLOpt._(138);

  /// TELOPT SSPI LOGON.
  static final sspiLogon = TLOpt._(139);

  /// TELOPT PRAGMA HEARTBEAT.
  static final pragmaHeartbeat = TLOpt._(140);

  /// Extended-Options-List.
  ///
  /// 参考：[RFC861](https://www.rfc-editor.org/rfc/rfc861.html)
  static final eol = TLOpt._(255);
}


lib/src/client.dart
part of "package:telnet/telnet.dart";

abstract class ITLConnectionTask {
  /// Telnet 客户端实例。
  ///
  /// 连接完成前，或者连接失败时，此属性均为 `null`，可通过判空来判断是否连接成功。
  ITelnetClient? get client;

  /// 连接结束时的回调事件。
  ///
  /// 其中 [client] 参数为 Telnet 客户端实例，若连接失败，则 [client] 为 `null`。
  void Function(ITelnetClient? client)? onDone;

  /// 连接过程中有异常发生时的回调。
  void Function(dynamic)? onError;

  /// 连接耗时。
  Duration get elapsed;

  /// 等待连接结束。
  Future<void> waitDone();

  /// 取消连接。
  ///
  /// 该方法只在连接进行中时有效，一旦连接结束（无论是连接成功还是失败）后再调用该方法则无任何效果。
  void cancel();
}

/// Telnet 连接任务类。
class TLConnectionTask implements ITLConnectionTask {
  TLConnectionTask._(Future<void> Function(TLConnectionTask) handler) {
    _watch.start();
    handler(this).catchError((error) {
      onError?.call(error);
    }).whenComplete(() {
      if (!_watch.isRunning) {
        _client?.terminate();
        _client = null;
      }
      _watch.stop();
      _task = null;
      onDone?.call(_client);
    });
  }

  ConnectionTask? _task;
  ITelnetClient? _client;
  final _watch = Stopwatch();

  @override
  ITelnetClient? get client => _client;

  @override
  void Function(ITelnetClient? client)? onDone;

  @override
  void Function(dynamic)? onError;

  @override
  Duration get elapsed => _watch.elapsed;

  @override
  Future<void> waitDone() async {
    while (_watch.isRunning) {
      await Future.delayed(const Duration(milliseconds: 5));
    }
  }

  @override
  void cancel() {
    _watch.stop();
    _task?.cancel();
    _task = null;
  }
}

abstract class ITelnetClient {
  /// 服务端的主机地址。
  String get remoteAddress;

  /// 服务端的端口号。
  int get remotePort;

  /// 交互过程中有异常发生时的回调。
  TLErrCallback? get onError;

  /// 连接被关闭时（断开连接或因发生异常而中止）的回调。
  TLDoneCallback? get onDone;

  /// 当有 Telnet 消息事件发生时（发送一个 Telnet 消息或接收到一个 Telnet 消息时）的回调。
  /// 可在这个回调中侦听和处理各种消息事件。
  TLEventCallback? get onEvent;

  /// 客户端的主机地址。
  String? get localAddress;

  /// 客户端的端口号。
  int? get localPort;

  /// 是否已建立连接。
  bool get isConnected;

  /// 中断连接。
  Future<void> terminate();

  /// 向服务端发送一个 Telnet 消息。
  ///
  /// 该消息可以是选项协商（[TLOptMsg]）、子选项协商（[TLSubMsg]），也可以是文本消息（[TLTextMsg]）。
  void write(TLMsg msg);

  /// 一次性向服务端发送多个 Telnet 消息。
  void writeAll(Iterable<TLMsg> messages);
}

/// Telnet 客户端类。
class TelnetClient implements ITelnetClient {
  /// 启动一个 Telnet 连接。
  ///
  /// 返回一个 [ITLConnectionTask] 实例，可使用该实例来取消连接、获取连接过程中发生的异常、获取连接耗时等。
  ///
  /// [timeout] 是指连接超时时间。
  ///
  /// [onError] 是在连接成功之后的交互过程中发生异常时的回调（例如异常断开连接），
  /// 正在连接的过程中所产生的异常请在 [ITLConnectionTask.onError] 回调中捕获。
  ///
  /// [onDone] 是连接被关闭（手动断开或由于网络异常等原因自动断开）时的回调。
  ///
  /// [onEvent] 是在有传输消息事件发生时（发送一个 Telnet 消息或接收到一个 Telnet 消息时）的回调，
  /// 可在这个回调中侦听和处理各种消息事件。
  static ITLConnectionTask startConnect({
    required String host,
    int port = 23,
    Duration timeout = const Duration(seconds: 10),
    TLErrCallback? onError,
    TLDoneCallback? onDone,
    TLEventCallback? onEvent,
  }) =>
      _getTask(host, port, timeout, onError, onDone, onEvent,
          (task) => RawSocket.startConnect(host, port));

  /// 启动一个安全的 Telnet 连接。
  ///
  /// 使用这个方法来创建的 Telnet 连接将使用 TLS 技术对传输的数据进行加密。
  /// 需要注意的是，必须服务端也支持 TLS 时才能正常使用。
  ///
  /// [securityContext]、[onBadCertificate] 和 [supportedProtocols] 参数的含义
  /// 请参阅 [RawSecureSocket.startConnect] 方法的文档注释。
  /// 其他参数含义请参阅 [TelnetClient.startConnect] 方法的文档注释。
  static ITLConnectionTask startSecureConnect({
    required String host,
    int port = 23,
    Duration timeout = const Duration(seconds: 10),
    SecurityContext? securityContext,
    bool Function(X509Certificate)? onBadCertificate,
    List<String>? supportedProtocols,
    TLErrCallback? onError,
    TLDoneCallback? onDone,
    TLEventCallback? onEvent,
  }) =>
      _getTask(
          host,
          port,
          timeout,
          onError,
          onDone,
          onEvent,
          (task) => RawSecureSocket.startConnect(host, port,
              context: securityContext,
              onBadCertificate: onBadCertificate,
              supportedProtocols: supportedProtocols));

  static ITLConnectionTask _getTask(
      String host,
      int port,
      Duration timeout,
      TLErrCallback? onError,
      TLDoneCallback? onDone,
      TLEventCallback? onEvent,
      Future<ConnectionTask> Function(ITLConnectionTask) f) {
    final task = TLConnectionTask._((task) async {
      task._task = await f(task).timeout(timeout - task.elapsed);
      if (task._task == null) {
        return;
      }
      final client = TelnetClient._(host, port, onError, onDone, onEvent);
      client._socket = await task._task!.socket.timeout(timeout - task.elapsed);
      if (client._socket == null) {
        return;
      }
      client._socket!.writeEventsEnabled = false;
      client._subscription = client._socket!.listen(client._onData,
          onError: client._onError, onDone: client._onDone);
      task._client = client;
    });
    task.onError = (err) => onError?.call(null, err);
    return task;
  }

  TelnetClient._(this.remoteAddress, this.remotePort, this.onError, this.onDone,
      this.onEvent);

  @override
  final String remoteAddress;

  @override
  late final int remotePort;

  @override
  final TLErrCallback? onError;

  @override
  final TLDoneCallback? onDone;

  @override
  final TLEventCallback? onEvent;

  RawSocket? _socket;
  StreamSubscription? _subscription;
  final _buffer = <int>[];
  _CodeState _state = _CodeState.normal;
  int _tick = 0;

  @override
  String? get localAddress => _socket?.address.address;

  @override
  int? get localPort => _socket?.port;

  @override
  bool get isConnected => _socket != null;

  @override
  Future<void> terminate() async {
    write(TLOptMsg(TLCmd.doIt, TLOpt.logout));

    final watch = Stopwatch()..start();
    while (watch.elapsedMilliseconds < 2000) {
      if (_socket == null) {
        watch.stop();
        return;
      }
      await Future.delayed(const Duration(milliseconds: 100));
    }

    watch.stop();
    await _terminateForce();
  }

  Future<void> _terminateForce() async {
    await _socket?.close();
    _socket = null;
  }

  @override
  void write(TLMsg msg) {
    final res = _socket?.write(msg.bytes);
    if (res != null && res > 0) {
      onEvent?.call(this, TLMsgEvent._(TLMsgEventType.write, msg));
    }
  }

  @override
  void writeAll(Iterable<TLMsg> messages) {
    final bytes = <int>[];
    for (final msg in messages) {
      bytes.addAll(msg.bytes);
    }
    final res = _socket?.write(bytes);
    if (res != null && res > 0) {
      for (final msg in messages) {
        onEvent?.call(this, TLMsgEvent._(TLMsgEventType.write, msg));
      }
    }
  }

  void _onData(RawSocketEvent event) {
    if (event != RawSocketEvent.read) {
      return;
    }

    final data = _socket?.read() ?? [];
    _tick++;
    for (final code in data) {
      _handleCharCode(code);
    }
    if (_state == _CodeState.normal) {
      final lastTick = _tick;
      Future.delayed(const Duration(milliseconds: 3), () {
        if (lastTick == _tick && _state == _CodeState.normal) {
          _outputMsg((data) => TLTextMsg.fromBytes(data));
        }
      });
    }
  }

  void _handleCharCode(int code) {
    switch (_state) {
      case _CodeState.normal:
        {
          if (code == TLCmd.iac.code) {
            _outputMsg((data) => TLTextMsg.fromBytes(data));
            _state = _CodeState.iac;
          }
          _buffer.add(code);
        }
        break;
      case _CodeState.iac:
        {
          _buffer.add(code);
          if (code == TLCmd.sb.code) {
            _state = _CodeState.subOpt;
          } else if (code == TLCmd.se.code) {
            _outputMsg((data) => TLSubMsg.fromBytes(data));
            _state = _CodeState.normal;
          } else {
            _state = _CodeState.opt;
          }
        }
        break;
      case _CodeState.opt:
        {
          _buffer.add(code);
          _outputMsg((data) => TLOptMsg.fromBytes(data));
          _state = _CodeState.normal;
        }
        break;
      case _CodeState.subOpt:
        {
          _buffer.add(code);
          if (code == TLCmd.iac.code) {
            _state = _CodeState.iac;
          }
        }
    }
  }

  void _outputMsg(TLMsg Function(List<int> data) msgBuilder) {
    final data = _buffer.take(_buffer.length).toList();
    _buffer.clear();
    if (data.isNotEmpty) {
      final msg = msgBuilder(data);
      onEvent?.call(this, TLMsgEvent._(TLMsgEventType.read, msg));

      if (msg is TLOptMsg && msg.opt == TLOpt.logout) {
        if (msg.cmd == TLCmd.doIt) {
          terminate();
        } else {
          _terminateForce();
        }
      }
    }
  }

  void _onDone() {
    _socket = null;
    _subscription?.cancel();
    _subscription = null;
    onDone?.call(this);
  }

  void _onError(dynamic error) {
    onError?.call(this, error);
  }
}


example/telnet_example.dart
import 'package:telnet/telnet.dart';

const host = "127.0.0.1";
const port = 23;
const username = "root";
const password = "admin";
const echoEnabled = true;

void main() async {
  // Create a Telnet connection task.
  final task = TelnetClient.startConnect(
    host: host,
    port: port,
    onEvent: _onEvent,
    onError: _onError,
    onDone: _onDone,
  );

  // Cancel the connection task.
  // task.cancel();

  // Wait the connection task finished.
  await task.waitDone();

  // Get the `TelnetClient` instance. It will be `null` if connect failed.
  final client = task.client;
  if (client == null) {
    print("Fail to connect to $host:$port");
  } else {
    print("Successfully connect to $host:$port");
  }

  await Future.delayed(const Duration(seconds: 10));

  // Close the Telnet connection.
  await client?.terminate();
}

var _hasLogin = false;
final _willReplyMap = <TLOpt, List<TLMsg>>{
  TLOpt.echo: [echoEnabled
      ? TLOptMsg(TLCmd.doIt, TLOpt.echo)                      // [IAC DO ECHO]
      : TLOptMsg(TLCmd.doNot, TLOpt.echo)],                   // [IAC DON'T ECHO]
  TLOpt.suppress: [TLOptMsg(TLCmd.doIt, TLOpt.suppress)],     // [IAC DO SUPPRESS_GO_AHEAD]
  TLOpt.logout: [],
};
final _doReplyMap = <TLOpt, List<TLMsg>>{
  TLOpt.echo: [echoEnabled
      ? TLOptMsg(TLCmd.will, TLOpt.echo)                      // [IAC WILL ECHO]
      : TLOptMsg(TLCmd.wont, TLOpt.echo)],                    // [IAC WONT ECHO]
  TLOpt.logout: [],
  TLOpt.tmlType: [
    TLOptMsg(TLCmd.will, TLOpt.tmlType),                      // [IAC WILL TERMINAL_TYPE]
    TLSubMsg(TLOpt.tmlType, [0x00, 0x41, 0x4E, 0x53, 0x49]),  // [IAC SB TERMINAL_TYPE IS ANSI IAC SE]
  ],
  TLOpt.windowSize: [
    TLOptMsg(TLCmd.will, TLOpt.windowSize),                   // [IAC WILL WINDOW_SIZE]
    TLSubMsg(TLOpt.windowSize, [0x00, 0x5A, 0x00, 0x18]),     // [IAC SB WINDOW_SIZE 90 24 IAC SE]
  ],
};

void _onEvent(TelnetClient? client, TLMsgEvent event) {
  if (event.type == TLMsgEventType.write) {
    print("[WRITE] ${event.msg}");

  } else if (event.type == TLMsgEventType.read) {
    print("[READ] ${event.msg}");

    if (event.msg is TLOptMsg) {
      final cmd = (event.msg as TLOptMsg).cmd; // Telnet Negotiation Command.
      final opt = (event.msg as TLOptMsg).opt; // Telnet Negotiation Option.

      if (cmd == TLCmd.wont) {
        // Write [IAC DO opt].
        client?.write(TLOptMsg(TLCmd.doNot, opt));
      } else if (cmd == TLCmd.doNot) {
        // Write [IAC WON'T opt].
        client?.write(TLOptMsg(TLCmd.wont, opt));
      } else if (cmd == TLCmd.will) {
        if (_willReplyMap.containsKey(opt)) {
          // Reply the option.
          for (var msg in _willReplyMap[opt]!) {
            client?.write(msg);
          }
        } else {
          // Write [IAC DON'T opt].
          client?.write(TLOptMsg(TLCmd.doNot, opt));
        }
      } else if (cmd == TLCmd.doIt) {
        // Reply the option.
        if (_doReplyMap.containsKey(opt)) {
          for (var msg in _doReplyMap[opt]!) {
            client?.write(msg);
          }
        } else {
          // Write [IAC WON'T opt].
          client?.write(TLOptMsg(TLCmd.wont, opt));
        }
      }

    } else if (!_hasLogin && event.msg is TLTextMsg) {

      final text = (event.msg as TLTextMsg).text.toLowerCase();
      if (text.contains("welcome")) {
        _hasLogin = true;
        print("[INFO] Login OK!");
      } else if (text.contains("login:") || text.contains("username:")) {
        // Write [username].
        client!.write(TLTextMsg("$username\r\n"));
      } else if (text.contains("password:")) {
        // Write [password].
        client!.write(TLTextMsg("$password\r\n"));
      }

    }
  }
}

void _onError(TelnetClient? client, dynamic error) {
  print("[ERROR] $error");
}

void _onDone(TelnetClient? client) {
  print("[DONE]");
}